### PR TITLE
refactor: Network test sync logic

### LINKED ISSUE #2747: Improve network test sync logic
The network sync logic could be improved by keeping a list of document heads that should be merged into target nodes. This will allow network events to happen in any order and also allow republish / retry logic to be added to the network implementation.

### PR DESCRIPTION
**Pull request by nasdf**
## Relevant issue(s)

Resolves #2747

## Description

This PR improves the network sync test logic to allow for further improvements to the network implementation.

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

`make test`

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

File: tests/integration/p2p.go
Comment (fredcarle) [inline tests/integration/p2p.go:330]: suggestion: The documentation says `waits for all selected nodes` but the function only receives one. You can probably do away with the loop and just get the node directly.

File: tests/integration/p2p.go
Comment (fredcarle) [inline tests/integration/p2p.go:354]: thougth: It's odd that we would be updating the expected heads only if already present in the actual heads. Shouldn't it be updated regardless?

File: tests/integration/p2p.go
Comment (nasdf) [reply tests/integration/p2p.go:330]: If `nodeID` has no value it will wait for all nodes.

File: tests/integration/p2p.go
Comment (nasdf) [reply tests/integration/p2p.go:354]: I believe the code is correct based on what we discussed regarding replicator target nodes sending updates of a document back to the replicator source node.

File: tests/integration/state.go
Comment (AndrewSisley) [inline tests/integration/state.go:58]: suggestion: Can we wrap these up in a dedicated type?  They are consuming a disproportionately large amount of attention on the state struct compared to their visibility needs.

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:1210]: thought: This is a significant change in the tests - and one that looks like it removes our ability to test concurrent P2P stuff (something that we very much want/need to test)?

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:1364]: question: Why is this only done using create-save, and not create-gql and the other create?  And why not on delete?

thought: This looks quite easy to mess up and forget, and it missing it might not create an obvious and immediate failure (just flaky tests).

**Review by AndrewSisley — CHANGES_REQUESTED** [review]
I've not finished reviewing, but want to block the merge until my concerns around the apparent loss of concurrency have been resolved.  The code looks simpler, but I think we may have lost the ability to test some important things.

**Comment by AndrewSisley** [comment]
todo: Please make sure you run the P2P tests a lot (IIRC Fred/I had to run them 100 or so times to reproduce the last solved issues) before merging this, we had flakiness problems with this before and they cost a lot of time.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:1210]: The function name is a bit misleading, but `Update` events are published immediately after a document is saved and shouldn't really change how we test things concurrently. 

We could also merge `waitForUpdateEvents` into `waitForSync` since the events are buffered if you would prefer that.

Comment (AndrewSisley) [reply tests/integration/utils2.go:1210]: Ah I see what you mean - it is not waiting for the sync, just the event publishing to the current node's event queue?

Why are you calling it then?

Comment (nasdf) [reply tests/integration/utils2.go:1210]: > Ah I see what you mean - it is not waiting for the sync, just the event publishing to the current node's event queue?

Yes

> Why are you calling it then?

Its a convenient way to track which doc heads are expected to be merged on remote nodes.

Comment (AndrewSisley) [reply tests/integration/utils2.go:1210]: > Its a convenient way to track which doc heads are expected to be merged on remote nodes.

Can you please document that really well?  Our integration tests are already super magic and far more complex than most, and that complexity is not really tested.  The P2P space is particularly complicated and prone to annoying, late-to-discover, flaky bugs and it is well worth over-explaining things to save future hassle.

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:770]: todo: This looks like it will break when collections are removed/deactivated?  Users currently can do this, although we likely have no P2P tests for it atm.

There may be other test things that currently break when doing that, but we should avoid adding new ones.

File: tests/integration/utils2.go
Comment (AndrewSisley) [inline tests/integration/utils2.go:760]: praise: I forgot we were currently relying on a sleep - thank you very much for removing this!

File: tests/integration/p2p.go
Comment (AndrewSisley) [inline tests/integration/p2p.go:340]: suggestion: Building on the other related documentation request, it might be worth highlighting that you mean publish to the local event queue, not to the pub/sub net or some other P2P related network thing.

File: tests/integration/utils2.go
Comment (shahzadlone) [reply tests/integration/utils2.go:1364]: +1 similar thought, for here and other instances of this change.

File: tests/integration/state.go
Comment (nasdf) [reply tests/integration/state.go]: Done

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:1185]: I added more documentation about how the document head are tracked. It should also be much easier to follow since its all moved to an `events.go` file. Let me know what you think.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:1391]: If I'm understanding this correctly the `createDoc` function calls the `create-save` and `create-gql` functions so they both should be covered.

File: tests/integration/state.go
Comment (nasdf) [reply tests/integration/state.go]: Done

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go]: This should be easy to handle. Just need to find which collections to deactivate from the p2p state.

- [ ] handle patchCollection

File: tests/integration/events.go
Comment (AndrewSisley) [inline tests/integration/events.go:36]: question: It looks like `p2pTopicEvents` is only ever 0 or 1, why is it not a boolean? Is this a bug/wip?

File: tests/integration/utils2.go
Comment (AndrewSisley) [reply tests/integration/utils2.go]: Changes look really good, thanks Keenan.  Has got me thinking about test Cid substitution too, in the future we might want to tweak `state.nodeP2P.DocHeads` to allow this.

File: tests/integration/utils2.go
Comment (AndrewSisley) [reply tests/integration/utils2.go:1393]: :grin: My bad, the github presentation of the diff helped me miss-read where this line was...  Ignore this thread :)

File: tests/integration/state.go
Comment (AndrewSisley) [inline tests/integration/state.go:28]: nitpick: `related testing state.`?

**Review by AndrewSisley — APPROVED** [review]
Looks a really nice improvement, thanks Keenan :)

File: tests/integration/utils2.go
Comment (islamaliev) [inline tests/integration/utils2.go:344]: praise: it's great that now waiting is done on the same thread. Earlier it used to call a goroutine for every `ConnectPeers` action and had subtle shortcomings.

File: tests/integration/utils2.go
Comment (islamaliev) [inline tests/integration/utils2.go:1429]: questions: this looks like unsolicited sync. Why is it unconditional?

**Review by islamaliev — APPROVED** [review]
Looks good. Core looks now much more intuitive. I just a have a question to clarify.

File: tests/integration/test_case.go
Comment (islamaliev) [inline tests/integration/test_case.go:351]: suggestion: why is it synced by default now? That resembles sequentially consistent memory order instead of relaxed one that we used before.
I think, if we want to sync, we should request it explicitly. Unsynced actions are closer to realistic behaviour.

Comment (AndrewSisley) [reply tests/integration/test_case.go:351]: todo: I think this is a misleading name and comment, IIRC `waitForUpdateEvents` does not wait for the P2P sync, but only the update event to be published to the event bus.  If that is correct, @nasdf can you please rework the property name and documentation to make that clearer.

Comment (islamaliev) [reply tests/integration/test_case.go:351]: what is this then if not wait: `<-s.nodeEvents[i].update.Message():`?

Comment (AndrewSisley) [reply tests/integration/test_case.go:351]: That is just the event bus, it is written to synchronously by the database when mutating documents - the P2P code also subscribes to this bus - so 'waiting' to the bus should never really block, and the P2P stuff hasn't even been sent over the network at that point.

If I understand it correctly.  I made the same comment as you did earlier for `waitForUpdateEvents` - I think Keenan improved the documentation and code location, but given that it is still tripping up reviewers, that too should probably be improved further.

File: tests/integration/utils2.go
Comment (nasdf) [reply tests/integration/utils2.go:1429]: There should always be an update event published to the local event bus if a document is updated or created.

File: tests/integration/state.go
Comment (nasdf) [reply tests/integration/state.go]: fixed

File: tests/integration/events.go
Comment (nasdf) [reply tests/integration/events.go]: Changed to bool

**Review by shahzadlone — APPROVED** [review]
Good Stuff 👍

**Pull request merged**
